master/internal/api_user.go

```diff
@@ -0,0 +1,151 @@
+package internal
+
+import (
+	"context"
+	"sort"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/determined-ai/determined/master/internal/db"
+	"github.com/determined-ai/determined/master/internal/grpc"
+	"github.com/determined-ai/determined/master/pkg/model"
+	"github.com/determined-ai/determined/proto/pkg/apiv1"
+	"github.com/determined-ai/determined/proto/pkg/userv1"
+)
+
+var errUserNotFound = status.Error(codes.NotFound, "user not found")
+
+func toProtoUserFromFullUser(user model.FullUser) *userv1.User {
+	var agentUserGroup *userv1.AgentUserGroup
+	if user.AgentUID.Valid || user.AgentGID.Valid || user.AgentUser.Valid || user.AgentGroup.Valid {
+		agentUserGroup = &userv1.AgentUserGroup{
+			AgentUid:   int32(user.AgentUID.ValueOrZero()),
+			AgentGid:   int32(user.AgentGID.ValueOrZero()),
+			AgentUser:  user.AgentUser.ValueOrZero(),
+			AgentGroup: user.AgentGroup.ValueOrZero(),
+		}
+	}
+	return &userv1.User{
+		Id:             int32(user.ID),
+		Username:       user.Username,
+		Admin:          user.Admin,
+		Active:         user.Active,
+		AgentUserGroup: agentUserGroup,
+	}
+}
+
+func getUser(d *db.PgDB, username string) (*userv1.User, error) {
+	user, err := d.UserByUsername(username)
+	switch {
+	case err == db.ErrNotFound:
+		return nil, errUserNotFound
+	case err != nil:
+		return nil, err
+	}
+	var protoAug *userv1.AgentUserGroup
+	agentUserGroup, err := d.AgentUserGroup(user.ID)
+	if agentUserGroup != nil {
+		protoAug = &userv1.AgentUserGroup{
+			AgentUid:   int32(agentUserGroup.UID),
+			AgentGid:   int32(agentUserGroup.GID),
+			AgentUser:  agentUserGroup.User,
+			AgentGroup: agentUserGroup.Group,
+		}
+	}
+	return &userv1.User{
+		Id:             int32(user.ID),
+		Username:       user.Username,
+		Admin:          user.Admin,
+		Active:         user.Active,
+		AgentUserGroup: protoAug,
+	}, err
+}
+
+func (a *apiServer) GetUsers(
+	context.Context, *apiv1.GetUsersRequest) (*apiv1.GetUsersResponse, error) {
+	users, err := a.m.db.UserList()
+	if err != nil {
+		return nil, err
+	}
+	result := &apiv1.GetUsersResponse{}
+	for _, user := range users {
+		result.Users = append(result.Users, toProtoUserFromFullUser(user))
+	}
+	sort.Slice(result.Users, func(i, j int) bool {
+		return result.Users[i].Username < result.Users[j].Username
+	})
+	return result, nil
+}
+
+func (a *apiServer) GetUser(
+	_ context.Context, req *apiv1.GetUserRequest) (*apiv1.GetUserResponse, error) {
+	fullUser, err := getUser(a.m.db, req.Username)
+	return &apiv1.GetUserResponse{User: fullUser}, err
+}
+
+func (a *apiServer) PostUser(
+	ctx context.Context, req *apiv1.PostUserRequest) (*apiv1.PostUserResponse, error) {
+	if err := grpc.ValidateRequest(
+		func() (bool, string) { return req.User != nil, "no user specified" },
```

Comment:
non-blocking: are we sure we don't to pass an error object to the request? That way we can have better typed errors potentially in the future?

Reply:
The grp error library takes strings as input, not errors